Fix `net_sfc_heating` bug for ACCESS-OM and add `mh_flux` diagnostic

rmholmes opened the pull request:
This PR:
- Adds diagnostics `mh_flux` and `total_ocean_mh_flux` which capture the heat flux due to melting ice in ACCESS-OM/ACCESS-CM.
- Adds a `total_ocean_liceht` diagnostic, being the globally-integrated `liceht` heat flux.
- Modifies the `net_sfc_heating` and `total_net_sfc_heating` diagnostics to include the heat fluxes associated with the ACCESS-OM/ACCESS-CM ice water fluxes `wfimelt`, `wfiform` and `licefw`.

This commit addresses ACCESS-OM2 issue https://github.com/COSIMA/access-om2/issues/139 and MOM5 issue https://github.com/mom-ocean/MOM5/issues/319

These changes only affect ACCESS-OM simulations.

My personal opinion is that the `net_sfc_heating` problem was a bug, and so I have not retained a legacy version as discussed at https://github.com/COSIMA/access-om2/issues/139. But I can add this if others think it would be useful?

I have checked these diagnostics by running two years of the `1deg_jra55_ryf` ACCESS-OM2 configuration and outputting all the surface heat flux diagnostics as well as some snapshots to check matching tendency (will comment with the results).

Several things I am unsure of/would like feedback on:
- I have not added a CICE diagnostic for `mh_flux` - should I?
- I have added the `total_net_sfc_heating` diagnostic. However, this does not include the heat input from frazil and so will not match the time-derivative of `total_ocean_heat` from snapshots. I tried adding a `total_ocean_frazil` diagnostic in `ocean_frazil` but it did not work (see https://github.com/rmholmes/MOM5/tree/total_ocean_frazil), likely because I'm not doing the global sum correctly. Is this desired? It may not be good for performance (@russfiedler?).
- I'm not sure if I should also be adding the `mh_flux` heat flux to the `latent` variable (that is later used for a few water-mass diagnostics)? E.g. it is not included here:
https://github.com/mom-ocean/MOM5/blob/4c3318ee9ab5b06530302d6a90c40b66e0ca798f/src/mom5/ocean_core/ocean_sbc.F90#L4170-L4174

Happy to recieve any/all feedback/suggestions. Once the changes are confirmed I can post confirming the relationships between the different variables.

rmholmes commented:
My check on a `1deg_jra55_ryf` two-year run confirms that the budgets/terms add up correctly. The following lists the total into ocean flux over a year (in W) and the sum-of-squares flux `sum(sum((area*flux)^2,'xt_ocean'),'yt_ocean')` for each variable as well as some residual calculations:
```
Total Flux = -3.8e+14 W, Sum of Squares = +1.4e+25 W^2, mh_flux (*)
Total Flux = -1.1e+14 W, Sum of Squares = +1.3e+24 W^2, fprec_melt_heat (*)
Total Flux = +0.0e+00 W, Sum of Squares = +0.0e+00 W^2, calving_melt_heat (*)
Total Flux = +5.8e+16 W, Sum of Squares = +7.5e+28 W^2, swflx (*)
Total Flux = -1.9e+16 W, Sum of Squares = +7.9e+27 W^2, lw_heat (*)
Total Flux = -3.7e+15 W, Sum of Squares = +4.6e+26 W^2, sens_heat (*)
Total Flux = -3.5e+16 W, Sum of Squares = +3.0e+28 W^2, evap_heat (*)
Total Flux = +0.0e+00 W, Sum of Squares = +0.0e+00 W^2, liceht (*)
Total Flux = -1.8e+14 W, Sum of Squares = +1.6e+25 W^2, sfc_hflux_pme (**, ***)
Total Flux = +9.1e+13 W, Sum of Squares = +3.2e+26 W^2, sfc_hflux_from_runoff (**)
Total Flux = +0.0e+00 W, Sum of Squares = +0.0e+00 W^2, sfc_hflux_from_calving (**)
Total Flux = -2.0e+14 W, Sum of Squares = +2.9e+27 W^2, net_sfc_heating
Total Flux = -1.1e+14 W, Sum of Squares = +2.6e+27 W^2, sfc_hflux_coupler
Total Flux = +3.7e+13 W, Sum of Squares = +3.1e+23 W^2, frazil_3d_int_z
Total Flux = +7.0e-02 W, Sum of Squares = +3.7e-02 W^2, sw_heat (vertically summed, ***)
Total Flux = -1.1e+14 W, Sum of Squares = +2.6e+27 W^2, temp_vdiffuse_sbc  (vertically summed, ***)
Total Flux = +9.1e+13 W, Sum of Squares = +3.2e+26 W^2, temp_rivermix  (vertically summed, ***)
Total Flux = -8.3e-02 W, Sum of Squares = +6.0e-01 W^2, sfc_hflux_coupler - sum(*)
Total Flux = +9.9e-02 W, Sum of Squares = +6.0e-01 W^2, net_sfc_heating - sfc_hflux_coupler - sum(**)
Total Flux = +1.5e-01 W, Sum of Squares = +1.1e-01 W^2, net_sfc_heating - sum(***)

Total Flux = -3.8e+14 W, total_ocean_mh_flux
Total Flux = -1.1e+14 W, total_ocean_fprec_melt_heat
Total Flux = +0.0e+00 W, total_ocean_calving_melt_heat
Total Flux = +5.8e+16 W, total_ocean_swflx
Total Flux = -1.9e+16 W, total_ocean_lw_heat
Total Flux = -3.7e+15 W, total_ocean_sens_heat
Total Flux = -3.5e+16 W, total_ocean_evap_heat
Total Flux = +0.0e+00 W, total_ocean_liceht
Total Flux = +9.1e+13 W, total_ocean_runoff_heat
Total Flux = +0.0e+00 W, total_ocean_calving_heat
Total Flux = +9.1e+13 W, total_ocean_river_heat
Total Flux = +1.1e+15 W, total_ocean_hflux_prec
Total Flux = -1.8e+14 W, total_ocean_hflux_pme
Total Flux = -1.2e+15 W, total_ocean_hflux_evap
Total Flux = -1.1e+14 W, total_ocean_hflux_coupler
Total Flux = -2.0e+14 W, total_net_sfc_heating

Total Flux = -1.6e+14 W, dHdt - calculated from snapshots
Total Flux = -2.0e+14 W, area-integral of net_sfc_heating 
Total Flux = +3.7e+13 W, area-integral of frazil_3d_int_z
Total Flux = +1.9e+07 W, total_net_sfc_heating - area-integral of net_sfc_heating
Total Flux = -1.1e+10 W, dHdt - area-integral of net_sfc_heating - area-integral of frazil_3d_int_z
```

The first set of numbers shows the 2D fluxes and shows that they add up as expected (see the last 3 lines of the first set which shows very small residuals, including comparing the `net_sfc_heating` to the vertically-summed 3D tendency diagnostics `temp_vdiffuse_sbc`, `temp_rivermix` and `sfc_hflux_pme`).

The second set of numbers are globally-integrated - these also match well. Although there are larger residuals when comparing to snapshots (likely associated with the way the sums are done), see last 2 lines.

rmholmes commented:
Just to confirm the variable relationships that I used above and that now work, the total heat flux into the ocean from surface forcing and ice-ocean exchanges is:

`net_sfc_heating + frazil_3d_int_z`

where,

`net_sfc_heating = sfc_hflux_coupler + sfc_hflux_pme + sfc_hflux_from_runoff + sfc_hflux_from_calving`

`sfc_hflux_coupler = swflx + lw_heat + fprec_melt_heat + calving_melt_heat + sens_heat + evap_heat + mh_flux + liceht`

Note the calving terms and liceht are currently zero in ACCESS-OM.

aekiss commented on src/mom5/ocean_core/ocean_sbc.F90:
```diff
@@ -1904,6 +1906,11 @@ subroutine ocean_sbc_diag_init(Time, Dens, T_prog)
+
+  id_mh_flux = register_diag_field('ocean_model','mh_flux', Grd%tracer_axes(1:2), &
+        Time%model_time, 'heat into ocean due to melting ice (>0 heats ocean)', &
+        '(W/M^2)', missing_value=missing_value,range=(/-1.e10,1.e10/),&
```
m should be lowercase in units

russfiedler commented:
@rmholmes  If running with the POP ice module (usually done for for ACCESS-OM)  you'll need to to add the global diagnostic in this block.
https://github.com/rmholmes/MOM5/blob/bfac193e92bfac11bad4ec1c0f51ed1d3bfd3ee9/src/mom5/ocean_tracers/ocean_frazil.F90#L538-L572

rmholmes commented:
That will only add the CICE diagnostic, is that correct?

There is a problem with the `total_ocean_frazil` diagnostic that I've added at lines 730-732 of that branch. It just spits out NaNs. Possibly I need to add masking (`*Grd%tmask(i,j,k)`)?

russfiedler commented:
If `pop_icediag` is .true. then you won't even perform the sum. There's a return at line 571. You'll just have some prefilled data in the file.
In other cases `T_diag(index_frazil)%field` appears to be initialised correctly to zero and the masking in its calculation looks fine.
Do you mean `_Fillvalue`s or `NaN`s in the output?  `NaN`s should never occur.

rmholmes commented:
Oh I see. I've added the global sum there. It compiles fine. I am testing it now.

rmholmes commented:
Yes that did it. I've now added the `total_ocean_frazil` diagnostic. The total heat flux into the ocean can now be obtained from `total_net_sfc_heating+total_ocean_frazil`.

Totals/residuals from my test are as follows:
```
Total Flux = -3.8e+14 W, total_ocean_mh_flux
Total Flux = -1.1e+14 W, total_ocean_fprec_melt_heat
Total Flux = +0.0e+00 W, total_ocean_calving_melt_heat
Total Flux = +5.8e+16 W, total_ocean_swflx
Total Flux = -1.9e+16 W, total_ocean_lw_heat
Total Flux = -3.7e+15 W, total_ocean_sens_heat
Total Flux = -3.5e+16 W, total_ocean_evap_heat
Total Flux = +0.0e+00 W, total_ocean_liceht
Total Flux = +9.1e+13 W, total_ocean_runoff_heat
Total Flux = +0.0e+00 W, total_ocean_calving_heat
Total Flux = +9.1e+13 W, total_ocean_river_heat
Total Flux = +1.1e+15 W, total_ocean_hflux_prec
Total Flux = -1.8e+14 W, total_ocean_hflux_pme
Total Flux = -1.2e+15 W, total_ocean_hflux_evap
Total Flux = -1.1e+14 W, total_ocean_hflux_coupler
Total Flux = +3.7e+13 W, total_ocean_frazil
Total Flux = -2.0e+14 W, total_net_sfc_heating

Total Flux = -1.6e+14 W, dHdt - calculated from snapshots
Total Flux = -2.0e+14 W, area-integral of net_sfc_heating 
Total Flux = +3.7e+13 W, area-integral of frazil_3d_int_z
Total Flux = +1.9e+07 W, total_net_sfc_heating - area-integral of net_sfc_heating
Total Flux = -3.4e+03 W, total_ocean_frazil - area-integral of frazil_3d_int_z
Total Flux = -1.1e+10 W, dHdt - area-integral of net_sfc_heating - area-integral of frazil_3d_int_z
Total Flux = -1.1e+10 W, dHdt - total_ocean_frazil - total_net_sfc_heating
```

I also fixed the W/m^2 captilization.

If everyone is happy that the `net_sfc_heating` diagnostic has changed (it is no longer wrong), then I think it is ok to merge? Note that I have not added `mh_flux` to `latent` for the WMT diagnostics (should I?).

aekiss commented:
great work, thanks @rmholmes!

aekiss commented:
@russfiedler does this look ok to you? I'm happy to merge it if you're happy with it.
